lib/gc/Transforms/GPU/GpuUtils.h

```diff
@@ -0,0 +1,241 @@
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#ifndef GPUUTILS_H
+#define GPUUTILS_H
+
+#include "gc/Utils/Log.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/DataLayoutInterfaces.h"
+
+#include <numeric>
+
+using namespace mlir;
+
+namespace mlir::gc {
+template <typename DerivedT> struct GpuPass {
+
+  int64_t getGpuPropertyAsInt(Builder &builder, StringRef name,
+                              int64_t defaultValue) {
+    if (auto mod = static_cast<DerivedT *>(this)
+                       ->getOperation()
+                       ->template getParentOfType<ModuleOp>()) {
+      DataLayout layout(mod);
+      if (auto value = layout.getDevicePropertyValue(
+              builder.getStringAttr("GPU" /* device ID*/),
+              builder.getStringAttr(name))) {
+        if (auto attr = dyn_cast<IntegerAttr>(*value)) {
+          return attr.getInt();
+        }
+      }
+    }
+    return defaultValue;
+  }
+
+  int64_t getNumEus(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "num_exec_units",
+                               static_cast<DerivedT *>(this)->numEus);
+  }
+
+  int64_t getNumEusPerSlice(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "num_exec_units_per_slice",
+                               static_cast<DerivedT *>(this)->numEusPerSlice);
+  }
+
+  int64_t getNumThreadsPerEu(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "num_threads_per_eu",
+                               static_cast<DerivedT *>(this)->numThreadsPerEu);
+  }
+
+  int64_t getCacheSize(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "L1_cache_size_in_bytes",
+                               static_cast<DerivedT *>(this)->cacheSize);
+  }
+
+  int64_t getVectorWidth(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "max_vector_op_width",
+                               static_cast<DerivedT *>(this)->vectorWidth);
+  }
+
+  int64_t getWorkGroupSize(Builder &builder) {
+    return getGpuPropertyAsInt(builder, "max_work_group_size",
+                               static_cast<DerivedT *>(this)->workGroupSize);
+  }
+};
+
+struct OpRewriter final : IRRewriter {
+  Location loc;
+
+  explicit OpRewriter(func::FuncOp &func)
+      : IRRewriter(func.getContext()), loc(func.getLoc()), func(func) {}
+
+  template <typename OpTy, typename... Args> OpTy create(Args &&...args) {
+    return RewriterBase::create<OpTy>(loc, std::forward<Args>(args)...);
+  }
+
+  arith::ConstantIndexOp getConstant(int64_t v) {
```

Comment:
Maybe could reuse `getValueOrCreateConstantIndexOp`

Reply:
getValueOrCreateConstantIndexOp has a different purpose, it converts OpFoldResult to a Value, or creates a new constant. This method returns an existing constant or creates a new one.

Reply:
The cse pass will merge the duplicate constant and we don't need to do it. For the `OpFoldResult`, you could create one easily from `int64_t` using `getAsIndexOpFoldResult(rewriter.getContext(), v)`.

Reply:
But I don't need OpFoldResult, I just need to create a new constant. getValueOrCreateConstantIndexOp() has a different purpose.

Reply:
I mean you could call `getValueOrCreateConstantIndexOp(getAsIndexOpFoldResult(rewriter.getContext(), v))` to get a constant from `int64_t` instead of writing one by yourself. `OpFoldResult` is a `ConstantAttribute` and `Value` union in MLIR.

Reply:
Why do we need OpFoldResult here? We just need to create a constant and this method is just a shorthand.
Which variant would you prefer?
```c++
getValueOrCreateConstantIndexOp(getAsIndexOpFoldResult(rewriter.getContext(), v));
// vs
rewriter.create<arith::ConstantIndexOp>(loc, v);
// vs
createConstant(v);
```
I'd definitely prefer the last one.

Reply:
Thanks for the explanation! I revisited your code and the current implementation is simple enough